src/main/java/org/jenkinsci/plugins/envinjectapi/util/EnvVarsResolver.java

```diff
@@ -0,0 +1,266 @@
+package org.jenkinsci.plugins.envinjectapi.util;
+
+import hudson.EnvVars;
+import hudson.FilePath;
+import hudson.Util;
+import hudson.model.AbstractBuild;
+import hudson.model.AbstractProject;
+import hudson.model.Action;
+import hudson.model.Cause;
+import hudson.model.CauseAction;
+import hudson.model.Computer;
+import hudson.model.Job;
+import hudson.model.Node;
+import hudson.model.Run;
+import hudson.slaves.EnvironmentVariablesNodeProperty;
+import hudson.slaves.NodeProperty;
+import hudson.slaves.NodePropertyDescriptor;
+import hudson.util.DescribableList;
+
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import jenkins.model.Jenkins;
+import jenkins.security.MasterToSlaveCallable;
+import org.jenkinsci.lib.envinject.EnvInjectException;
+import static org.jenkinsci.plugins.envinjectapi.util.CauseHelper.ENV_CAUSE;
+import static org.jenkinsci.plugins.envinjectapi.util.CauseHelper.ENV_ROOT_CAUSE;
+
+/**
+ * Provides utility methods for resolving environment variables.
+ * @author Gregory Boissinot
+ * @author Oleg Nenashev
+ */
+public class EnvVarsResolver {
```

Comment:
It's horrifying but I'll let it go because there is no net increase in horrify-ingness.  Three out of 4 Elder Gods approved this code.